TI-Superstructure

larswik opened the pull request.

larswik committed:
Started withy TI-Superstructure

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Cross Frames" UniqueId="370c7087-2b8b-4134-bbb2-69bc53b0fbe5">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Spatial_Container_1XtGZjG59LPCtYlqeIWvl" />
+					<Items>
+						<DocTemplateItem UniqueId="3548debb-c326-4e23-9a10-142ab59e5819" RuleParameters="Type=IfcElementAssembly;">
+							<Documentation>CROSS_BRACING?</Documentation>
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Floor Beams" UniqueId="abd97f83-66c8-4897-900c-969f879bf985">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Spatial_Container_1XtGZjG59LPCtYlqeIWvl" />
+					<Items>
+						<DocTemplateItem UniqueId="791590a9-c5a0-4842-9c42-a4d3f3bef19b" RuleParameters="Type=IfcBeam;">
+							<Documentation>JOIST?</Documentation>
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Stringers" UniqueId="a74dd762-c3dc-49f1-a375-d9863cfb449b">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Spatial_Container_1XtGZjG59LPCtYlqeIWvl" />
+					<Items>
+						<DocTemplateItem UniqueId="6497f1e1-a970-4457-9f2f-6ca4ebb2a91a" RuleParameters="Type=IfcMember;">
+							<Documentation>STRINGER?</Documentation>
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
```
Are '?' in lines 113, 121 and 129 intentional? Were you trying to steer a discussion?

larswik replied:
> Are '?' in lines 113, 121 and 129 intentional? Were you trying to steer a discussion?

They will be gone after I get the IfcDoc update from @SergejMuhic.

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Girders" UniqueId="148dafa3-aa4c-406b-85fd-c557ecbe3020">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Spatial_Container_1XtGZjG59LPCtYlqeIWvl" />
+					<Items>
+						<DocTemplateItem UniqueId="47c95fad-26c1-417f-bd68-0e2eeb9b18df" RuleParameters="Type=IfcElementAssembly;">
+							<Documentation>GIRDER</Documentation>
+							<Concepts>
+								<DocTemplateUsage Name="Type" UniqueId="03736d66-3828-43db-be0a-5feb72cae558">
+									<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Object_Predefined_Type_0b989XcDzD2fDaLbDNNAGX" />
+								</DocTemplateUsage>
+							</Concepts>
+						</DocTemplateItem>
+					</Items>
+				</DocTemplateUsage>
```
The inner most `DocTemplateUsage` is referencing `Object_Predefined_Type`, however no rule parameters were defined. Intentional?

larswik replied:
> The inner most `DocTemplateUsage` is referencing `Object_Predefined_Type`, however no rule parameters were defined. Intentional?

Waiting for IfcDoc update...

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						<DocTemplateItem UniqueId="c317cfeb-cf7a-47f5-8808-2be63670b4ea" RuleParameters="Type=IfcElementAssembly;">
+							<Documentation>TRUSS</Documentation>
+							<Concepts>
+								<DocTemplateUsage Name="Type" UniqueId="db3db62a-f865-4197-b084-c08b9b431e6a">
+									<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Object_Predefined_Type_0b989XcDzD2fDaLbDNNAGX" />
+									<Items>
+										<DocTemplateItem UniqueId="852a63c7-0de9-4c68-9415-870710c8eef8" RuleParameters="TypePredefinedType=TRUSS;" />
+									</Items>
+								</DocTemplateUsage>
```
I am not convinced about the `TypePredefinedType` rule id: is the intention to only apply to the connected `IfcTypeObject` (i.e. `IfcElementAssemblyType.PredefinedType`)?

larswik replied:
Strange indeed. Didn't show up in the IfcDoc UI? Should be fixed now  with https://github.com/bSI-InfraRoom/TI-Bridge/pull/9/commits/11bd24d5859e92952dd7d8ff99bec3e6b6b7a95f

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						<DocTemplateItem UniqueId="b1b1692b-afe7-4023-803e-35073b0dd138" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Girder Segment Left" UniqueId="47cc5512-8b54-4f63-8d87-4cc95c712ddf">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Element_Connectivity_0Nv2C1Ibb9HROhB6JsocmI" />
+					<Items>
+						<DocTemplateItem UniqueId="cebb616d-eddd-498c-b1ac-5a843f691c7a" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Girder Segment Right" UniqueId="042e7da0-2c21-4c9d-8542-b9acddd03b46">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Element_Connectivity_0Nv2C1Ibb9HROhB6JsocmI" />
+					<Items>
+						<DocTemplateItem UniqueId="19632d3f-bf1a-4c96-9f50-305415fc1276" />
+					</Items>
+				</DocTemplateUsage>
```
Looking forward to how this will turn out.

pjanck commented on IFC4x3/Templates/Object Connectivity/Spatial Structure/Spatial Container/Spatial Container PredefinedType/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,23 @@
+					<Rules>
+						<DocModelRuleAttribute Name="RelatedElements" Identification="Type">
+							<Rules>
+								<DocModelRuleEntity Name="IfcProduct">
+									<References>
+										<DocTemplateDefinition xsi:nil="true" href="Object_Predefined_Type_0b989XcDzD2fDaLbDNNAGX" />
+									</References>
+								</DocModelRuleEntity>
```
The `Reference` would need an `IdPrefix='Type'` to use `TypePredefinedType` as a rule id, I believe.

larswik replied:
This needs a bit of elaboration so I understand what to do exactly.

larswik replied:
Ping @SergejMuhic and @pjanck: Can we have a look at this together?

pjanck reviewed (changes requested):
I would greet filled in `documentation` XML nodes as the case in other PRs.

SergejMuhic commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+<?xml version="1.0" encoding="utf-8"?>
+<DocModelView xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="TI-Bridge-Superstructure" UniqueId="fcc6bbbc-8f8f-40a2-96d2-a00b1d46ea4c">
+	<ConceptRoots>
+		<DocConceptRoot Name="Girder Segment" UniqueId="5d25181b-c624-45c3-b4b2-16f9f5e8fede">
```
I see that we should probably bring this one either together with the showcase from AASHTO or remove the other one completely.

larswik replied:
> I see that we should probably bring this one either together with the showcase from AASHTO or remove the other one completely.

How do we decide that?

SergejMuhic commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+			<ApplicableEntity xsi:type="DocEntity" xsi:nil="true" href="IfcBeam" />
+			<Concepts>
+				<DocTemplateUsage Name="Name" UniqueId="0faf1f26-6ecf-49ae-bb15-ffced2edb174">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Object_User_Identity_31di66dFr7AJKdxDG2D1A" />
```
Would be great if you added the descriptions from the RA report. I will wait for those to not have to switch back and forth between the RA report and the PR.

larswik replied:
Done.

SergejMuhic commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Connection Head" UniqueId="b04f73f2-4e19-433e-8152-c70c0577876c">
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Path_Connectivity_1G5mELXyP9BgRydr_42n5F" />
```
Have to start thinking about adding the values, otherwise these are more usages for the AbRV or Bridge RV or whatever MVD it would turn out to be.

larswik replied:
I needed a structure before thinking of adding values.

larswik replied:
A bit unclear to me how to specify e.g. Name when there are more than one instance, e.g. "Girder Segment". Is seems to be possible to specify e.g. regular_expression. Possible in IfcDoc (e.g. "'reg' ^GS_ ;)"?

larswik replied:
I will add names for elements and materials (wherever I can put in a value).

SergejMuhic commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						<DocTemplateItem UniqueId="c9c75962-4f57-4abe-b9eb-321b88e06c3f" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Alignments" UniqueId="4dc839ff-3200-49fd-b765-8426124d3a3f">
```
I would propose to re-use this. I have put some thought into it already. See https://github.com/bSI-InfraRoom/TI-Bridge/blob/ae9381221f18493e005fb18cfcf5f02f27ea02fe/IFC4x3/ModelViews/TI-Bridge/DocModelView.xml#L60

larswik replied:
Partly resolved using CT "Axis Offset Curve". Some hacking still required.

SergejMuhic commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,283 @@
+						<DocTemplateItem UniqueId="d0bb99f6-8e52-4cb6-82e3-2e45b606841c" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Alignment" UniqueId="5e82185b-39f1-457c-9dcd-09c34a4772d2">
```
Maybe we need to revise the template. I will add the ruleIDs update to your repo so that you can access nested template ruleIDs.

larswik replied:
We would need a station template from BuildingSMART.

larswik replied:
> Maybe we need to revise the template. I will add the ruleIDs update to your repo so that you can access nested template ruleIDs.

We need a station CT from BuildingSMART for this.

SergejMuhic replied:
Since the template needs to go back to the alignment as it is an alignment reference, I think we need a new basis. I have compiled a proposal in the alignment instruction if you look at the Bridge root. I will merge and reassign. No need to do anything in this PR then.

The information on station has to be modelled in a template however.

larswik committed:
Update DocModelView.xml

Added descriptions and empty column definitions (requirements) to Diaphragm.

larswik committed:
Some more work on Superstructure Table definition

Had to restart a couple of times due to an IfcDoc crash. Restarted the machine and refreshed git (now ok again).

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,355 @@
+				<DocTemplateUsage Name="Material" UniqueId="a23c9639-1794-468c-875f-3621a7b4576c">
+					<Documentation>Material of the girder segment.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" Name="Material" UniqueId="05b10ba1-5fb2-4d43-ac50-875fc9c85e58" />
+				</DocTemplateUsage>
```
```suggestion
				<DocTemplateUsage Name="Material" UniqueId="a23c9639-1794-468c-875f-3621a7b4576c">
					<Documentation>Material of the diaphragm.</Documentation>
					<Definition xsi:type="DocTemplateDefinition" Name="Material" UniqueId="05b10ba1-5fb2-4d43-ac50-875fc9c85e58" />
				</DocTemplateUsage>
```

larswik committed:
As far as I can get on TI Superstructure for now.

See separate comment for remaining issues.

larswik commented:
Remaining issues for now (in addition to open comments above):
- Superstructure-alignments?
- Element Decomposition Predefined Type - Can't select IfcBeam/PredefinedType - list has predefined types for root (e.g. IfcElementAssembly)
- Concrete parameters (names, materials, stations, ...)
- Girder Segment
	- Connection head/tail/support
		- Element Connectivity With Realizing ELement (new CT)
		- Element Connectivity With Realizing Element with Predefined Type?
		- Connection support connecting to "substructure"...?
	- Element Connectivity With Realizing Element (new CT for Girder Segment - Connection Support)
	- Reinforcing embedded - IfcReinforcingBar (Decomposition?)
	- Tendons embedded - IfcTendon (Decomposition?)
	- Stiffeners - IfcPlate/STIFFENER_PLATE? (Decomposition?)
	- Camber (the camber of a beam may be defined by ASSIGNING a StructuralCurveMember with displacement coordinates?)
- How do we specify more than one instance? Do we need tables on the side specifying parameters?
- Cross frame
	- How to specify girder segment left/right? (ConnectedFrom/ConnectedTo)
	- How to specialize for IfcReferent and Station?
- Diaphragm
	- Connection head/tail (How to be able to specify ConnectedTo => IfcBeam/GIRDER_SEGMENT?)
	- Connection support connecting to "substructure"...?
	- Reinforcing (Decomposition?)
	- Tendons (Decomposition?)
- Truss
	- Element decomposition predefined type (see above)
	- Connection to "substructure"?

larswik committed:
Update DocModelView.xml

larswik committed:
Update DocModelView.xml

larswik committed:
Implemented as far as I can

Remaining issues see separate PDF.

larswik commented:
Remaining issues in attached document. 
[TI Bridge - Superstructure.pdf](https://github.com/bSI-InfraRoom/TI-Bridge/files/12714329/TI.Bridge.-.Superstructure.pdf)

larswik committed:
Update DocModelView.xml

larswik commented:
Update 2023-10-12 - remaining main issues:
- Superstructure => Alignments
- Girder => Axis offset curve
- Girder segment => Camber
- Diaphragm => Connection support (connecting diaphragm to superstructure)
- Truss => Connections (connecting truss to other superstructure elements)

pjanck commented on IFC4x3/Templates/Object Connectivity/Element Connectivity/Element Connectivity with Predefined Type/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,24 @@
+		<DocModelRuleAttribute Name="ConnectedFrom" />
+		<DocModelRuleAttribute Name="ConnectedTo">
+			<Rules>
+				<DocModelRuleEntity Name="IfcRelConnectsElements" Identification="RelatedElement">
+					<Rules>
+						<DocModelRuleAttribute Name="RelatedElement" Identification="RelatedElement">
```
Both rules have the same `Identification`, which will lead to confusion when writing rules. Propose to remove the one on `DocModelRuleEntity`.

larswik committed:
Changes after consultation with Sergej

- Girders for Superstructure can be both IfcElementAssembly/GIRDER and IfcBeam/GIRDER_SEGMENT
- Requirement Girder.Axis handled through Girder Segment.Solid Geometry
- Requirement Girder Segment.Camber handled through CT Structural Curve Assignment (new CT - check place in the structure)

pjanck commented on IFC4x3/Templates/Object Connectivity/Element Connectivity/Element Connectivity with Predefined Type/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,24 @@
+			<Rules>
+				<DocModelRuleEntity Name="IfcRelConnectsElements">
+					<Rules>
+						<DocModelRuleAttribute Name="RelatedElement" Identification="RelatedElement">
+							<Rules>
+								<DocModelRuleEntity Name="IfcElement">
+									<References>
+										<DocTemplateDefinition xsi:nil="true" href="Object_Predefined_Type_0b989XcDzD2fDaLbDNNAGX" />
```
Propose to add a prefix to the `Reference` node, e.g. `RelatedElement`. This enables easier uniqueness of RuleIDs while checking.

pjanck commented on IFC4x3/Templates/Structural Curve Assignment/DocTemplateDefinition.xml:
This whole ConceptTemplate, while fairly in-depth, only defines one `Identification`, thus only one setting can be checked for. Is this intentional?

larswik replied:
> This whole ConceptTemplate, while fairly in-depth, only defines one `Identification`, thus only one setting can be checked for. Is this intentional?

We'll have to check this. I tried to emulate a CT that @SergejMuhic sent me as an image. I might have missed important information.

larswik replied:
We have no more requirements for this so can't do more at this stage.

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,473 @@
+						<DocTemplateItem UniqueId="700e477d-19c6-428d-bf8e-1e8a22348e53" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Segments" UniqueId="aaef6f97-f84c-4391-a290-c5e4e6021bb2">
+					<Documentation>Segmented girders may be decomposed into segments for each continuous section.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Element_Decomposition_with_Predefined_Type_2doY2TqVr60xar2GKSiGHH" />
+					<Items>
+						<DocTemplateItem UniqueId="67edb96a-b5e3-4748-8178-38815481a090" RuleParameters="RelatedObjects=IfcBeam;PredefinedType=GIRDER_SEGMENT;">
+							<Documentation>GIRDER_SEGMENT</Documentation>
+						</DocTemplateItem>
+						<DocTemplateItem UniqueId="6a71ad28-e36e-41aa-a8ff-33954c701332" />
+					</Items>
+				</DocTemplateUsage>
```
There is an item without any `RuleParameters` in line 438. Intentional?

larswik replied:
> There is an item without any `RuleParameters` in line 438. Intentional?

It seems like IfcDoc sometimes adds these "empty items". Maybe I need to go through and manually remove these?

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,473 @@
+						<DocTemplateItem UniqueId="f2a7cee8-d90d-47d7-933a-2fe605a9a997" RuleParameters="TypePredefinedType=GIRDER;" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Material" UniqueId="b1809148-befc-43d4-a8f8-f35448334073">
+					<Documentation>Common material that applies to all segments of the girder.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Material_Single_1VEQWKHA1Fiut6tQe5OFnK" />
+					<Items>
+						<DocTemplateItem UniqueId="43226abe-20a0-49eb-b9fc-45ae97acd961" />
+						<DocTemplateItem UniqueId="700e477d-19c6-428d-bf8e-1e8a22348e53" />
+					</Items>
+				</DocTemplateUsage>
```
There are items without any `RuleParameters` in lines 427 and 428. Intentional?

larswik replied:
> There are items without any `RuleParameters` in lines 427 and 428. Intentional?

See response to previous comment.

pjanck commented on IFC4x3/ModelViews/TI-Bridge-Superstructure/DocModelView.xml:
```diff
@@ -0,0 +1,473 @@
+						<DocTemplateItem UniqueId="09aff01a-5e28-4f0f-a475-1f37e0b15fe1" RuleParameters="PositioningElement=IfcAlignment;" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Axis" UniqueId="c923e180-fc6c-494a-b859-db1be1bfa646">
+					<Documentation>Axis curve defined as a sub-span with offsets relative to the alignment curve. This requirement is handled by Girder Segment.Solid Geometry. The girders of the superstructure may be realized either by IfcElementAssembly/GIRDER or IfcBeam/GIRDER_SEGMENT. IfcBeam/GIRDER_SEGMENT has a geometric representation which is of type Body SectionedSolidHorizontal.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" Name="Axis" UniqueId="dc803a94-5992-4c3e-8e8d-38051cd52669" Type="IfcElementAssembly" />
+				</DocTemplateUsage>
```
This seems incomplete. Please double check. (It might be ok - I'm curious how the mvdXML will look like here.)

larswik replied:
> This seems incomplete. Please double check. (It might be ok - I'm curious how the mvdXML will look like here.)

We need a station CT from BuildingSMART for this.

pjanck reviewed (changes requested):
I haven't looked through the whole `DocModelView`, but let's first resolve these comments before I continue.

larswik committed:
Updates after walkthrough with @SergejMuhic
